pkg/controller/component/synthesize_component.go

```diff
@@ -210,13 +211,30 @@ func buildSynthesizedComponent(reqCtx intctrlutil.RequestCtx,
 	// replace podSpec containers env default credential placeholder
 	replaceContainerPlaceholderTokens(synthesizeComp, GetEnvReplacementMapForConnCredential(synthesizeComp.ClusterName))
 
-	// replace podSpec containers env component connection credential placeholder
-	// TODO(xingran): This is a temporary solution used to reference component connection credentials defined in ComponentDefinition. it will be refactored in the future.
-	replaceContainerPlaceholderTokens(synthesizeComp, GetEnvReplacementMapForCompConnCredential(synthesizeComp.ClusterName, synthesizeComp.Name))
-
 	return synthesizeComp, nil
 }
 
+func buildComp2CompDefs(cluster *appsv1alpha1.Cluster, clusterCompSpec *appsv1alpha1.ClusterComponentSpec) map[string]string {
+	if cluster == nil {
+		return nil
+	}
+	if len(cluster.Spec.ComponentSpecs) == 0 {
+		if clusterCompSpec == nil || len(clusterCompSpec.ComponentDefRef) == 0 {
+			return nil
+		}
+		return map[string]string{
+			clusterCompSpec.Name: clusterCompSpec.ComponentDefRef,
+		}
+	}
+	mapping := make(map[string]string)
+	for _, comp := range cluster.Spec.ComponentSpecs {
+		if len(comp.ComponentDefRef) > 0 {
+			mapping[comp.Name] = comp.ComponentDefRef
```

Comment:
Should this be comp.ComponentDef? componentDefRef is the reference of clusterDefinition.Spec.componentDefs[x].name

Reply:
fixed